src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,1273 @@
+                    gates.ResetChannel(0, [1, 0])
+                )  # ResetChannel(qubit, [prob0, prob1])
+                circ.add(gates.H(0))
+                circ.add(gates.S(0))
+                circ.add(
+                    gates.Unitary(
+                        BasisOps_13[idx_basis_ops],
+                        0,
+                        trainable=False,
+                        name="%s" % (gatenames[idx_basis_ops]),
+                    )
+                )
+
+            elif idx_basis_ops == 12:  # Reset, prepare |0> state
+                circ.add(
+                    gates.ResetChannel(0, [1, 0])
+                )  # ResetChannel(qubit, [prob0, prob1])
+                circ.add(
+                    gates.Unitary(
+                        BasisOps_13[idx_basis_ops],
+                        0,
+                        trainable=False,
+                        name="%s" % (gatenames[idx_basis_ops]),
+                    )
+                )
+
+            column_of_js = np.zeros((4**nqubits, 1))
+
+            # ================= START OF 4 MEASUREMENT BASES =================|||
+            for j in range(0, 4**nqubits):
+                new_circ = measure_1qb_state_new(j, circ)
+                # expectation_val = GST_measure_13(new_circ, idx_basis_ops, k, j, nshots, save_data)
+                expectation_val = GST_measure(
+                    new_circ,
+                    k,
+                    j,
+                    nshots,
+                    nqubits,
+                    class_qubit_gate,
+                    theta,
+                    idx_basis_ops,
+                    save_data,
+                )
+                column_of_js[j, :] = expectation_val
+
+            Bjk_tilde_1qb[idx_basis_ops, :, k] = column_of_js.reshape(
+                4,
+            )
+        idx_basis_ops += 1
+
+    toc = time.time() - tic
+    print("13 basis operations tomography done. %.4f seconds\n" % (toc))
+
+    #######################################################################
+    ### Get noisy Pauli Transfer Matrix for each of 13 basis operations ###
+    #######################################################################
+
+    # T = np.matrix([[1,1,1,1],[0,0,1,0],[0,0,0,1],[1,-1,0,0]])
+    T = np.array([[1, 1, 1, 1], [0, 0, 1, 0], [0, 0, 0, 1], [1, -1, 0, 0]])
+    Bjk_hat_1qb = np.zeros((13, 4**nqubits, 4**nqubits))
+    for idx_BO in range(0, 13):
+        Bjk_hat_1qb[idx_BO, :, :] = (
+            T @ np.linalg.inv(gjk_1qb) @ Bjk_tilde_1qb[idx_BO, :, :] @ np.linalg.inv(T)
+        )
+
+    # Reshape
+    Bjk_hat_1qb_reshaped = np.zeros((16, 13))
+    for idx_basis_ops in range(0, 13):
+        temp = np.reshape(
+            Bjk_hat_1qb[idx_basis_ops, :, :], [16, 1], order="F"
+        )  # 'F' so that it stacks columns.
+        Bjk_hat_1qb_reshaped[:, idx_basis_ops] = temp[:, 0].flatten()
+
+    return Bjk_hat_1qb, Bjk_hat_1qb_reshaped, BasisOps_13
+
+
+def GST_2qb_basis_operations(
```

Comment:
I think this function needs to be rewritten from scratches. A lot of code repetitions could be avoided by defining general helper functions that could be also used in the 1qb_basis_operations. In general, follow the suggestions I made above and apply them here as well.

Reply:
I've merged ```GST_1qb_basis_operations``` and ```GST_2qb_basis_operations``` into a single function. I'm clear on the helper functions for state preparation and measurements, which was reflected in this iteration of the code. But I'm not clear how to write the helper functions for 13 basis operations and 241 operations.